### PR TITLE
[clang] Fix CTAD not respect default template arguments that were added after the definition.

### LINKED ISSUE #69987: Default template arguments added in later redeclaration not considered for CTAD
```cpp
template<class> struct X {};
template<class=void> struct X;

X x;
```

Accepted by GCC and MSVC, rejected by Clang:

```
<source>:4:3: error: no viable constructor or deduction guide for deduction of template arguments of 'X'
    4 | X x;
      |   ^
<source>:1:24: note: candidate template ignored: couldn't infer template argument ''
    1 | template<class> struct X {};
      |                        ^
<source>:1:24: note: candidate function template not viable: requires 1 argument, but 0 were provided
    1 | template<class> struct X {};
      |                        ^
1 error generated.
Compiler returned: 1
```

This affects `<memory_resource>` in GCC 13's libstdc++, which defines `polymorphic_allocator` first before adding a default template argument in a later declaration.

### PR DESCRIPTION
**Pull request by hokein**
Fixes https://github.com/llvm/llvm-project/issues/69987

### CODE REVIEW

File: clang/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -685,6 +685,8 @@ Bug Fixes in This Version
   (`#62157 <https://github.com/llvm/llvm-project/issues/62157>`_) and
   (`#64885 <https://github.com/llvm/llvm-project/issues/64885>`_) and
   (`#65568 <https://github.com/llvm/llvm-project/issues/65568>`_)
+- Fix an issue where clang doesn't respect detault template arguments that
+  are added in a later redeclaration for CTAD. (#69987 <https://github.com/llvm/llvm-project/issues/69987>`_)
```
Comment (Fznamznon) [inline clang/docs/ReleaseNotes.rst:689]: A little nit to align the style
```suggestion
  are added in a later redeclaration for CTAD.
  Fixes (#69987 <https://github.com/llvm/llvm-project/issues/69987>`_)
```

Comment (hokein) [reply clang/docs/ReleaseNotes.rst]: Done.

**Review by cor3ntin — APPROVED** [review]
LGTM, thanks

**Comment by hokein** [comment]
Thanks for the review.

**Comment by shafik** [comment]
Can we please add more detailed summaries, we want the git log to be sufficient for triaging various issues without having to dig into each commit individually.

File: clang/lib/Sema/SemaTemplate.cpp [hunk 2]
```diff
@@ -2061,13 +2070,13 @@ DeclResult Sema::CheckClassTemplate(
...
-          PrevClassTemplate
-              ? PrevClassTemplate->getMostRecentDecl()->getTemplateParameters()
-              : nullptr,
+          PrevClassTemplate ? GetTemplateParameterList(PrevClassTemplate)
+                            : nullptr,
           (SS.isSet() && SemanticContext && SemanticContext->isRecord() &&
            SemanticContext->isDependentContext())
               ? TPC_ClassTemplateMember
-              : TUK == TUK_Friend ? TPC_FriendClassTemplate : TPC_ClassTemplate,
+          : TUK == TUK_Friend ? TPC_FriendClassTemplate
```
Comment (shafik) [inline clang/lib/Sema/SemaTemplate.cpp:2078]: This looks like an unrelated change and it looks funny.
